make: Add ineffassign as a part of precheck

vadorovsky opened the pull request:
Fixes #3609

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4374)
<!-- Reviewable:end -->

joestringer commented on pkg/identity/identity.go:
```diff
@@ -127,7 +127,7 @@ func (pair *IPIdentityPair) IsHost() bool {
 	var suffix string
 	if !pair.IsHost() {
-		ones := net.IPv6len
+		var ones int
```
This case looks interesting to me actually, there's two broad cases: Mask is nil, or not... then two inner cases: IPv4 or IPv6. It looks like if it's not IPv4, we could end up dereferencing a nil Mask.

Perhaps it should be something more like:

```
var ones int
if pair.Mask == nil {
    if pair.IP.To4() != nil {
        ones = net.IPv4len
    } else {
        ones = net.IPv6len
    }
} else {
    ones, _ = pair.Mask.Size()
}
```

vadorovsky replied:
Yes, that looks reasonable. I also had a feeling that the if statement in the current shape is a little bit unclear.

joestringer commented on pkg/kvstore/etcd.go:
```diff
@@ -580,7 +580,7 @@ reList:
 
 					// Retrieve latest revision
-					lastRev = 0
+					// lastRev = 0
```
It looks to me like this should be reused at the start of the `reList` loop, so it's not really ineffassign. Maybe we can find a way to teach ineffassign about this?

vadorovsky replied:
I think it will be hard to teach ineffassign about this (it just doesn't have a support of skipping errors). But I have an another idea - we can wrap `ineffassign` call in `contrib/scripts/check-ineffassign.sh` and do there something like:

```
ineffassign $TOPDIR | grep -v "# ineffassign: ignore"
````

Because ineffassign output shows the whole line.

joestringer commented on pkg/workloads/containerd.go:
```diff
@@ -106,7 +106,7 @@ func newContainerDClient(opts workloadRuntimeOpts) (WorkloadRuntime, error) {
 		ep = "unix://" + ep
 	}
 	rsc, err := newCRIClient(context.WithValue(context.Background(), epOpt, ep))
-	return &containerDClient{c, rsc}, nil
```
Nice catch.

joestringer commented on test/runtime/Policies.go:
```diff
@@ -1562,7 +1562,7 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
 		Expect(httpd1EndpointModel.Status.Policy).To(Not(BeNil()), "Expected non-nil policy for endpoint %s", helpers.Httpd1)
 
-		httpd1SecurityIdentity = httpd1EndpointModel.Status.Identity.ID
+		// httpd1SecurityIdentity = httpd1EndpointModel.Status.Identity.ID
```
Why is this redundant but the following line isn't?

vadorovsky replied:
Because `httpd2SecurityIdentity` variable is used in line 1569. `httpd1...` variable isn't used by the rest of the function at all.

joestringer reviewed (commented):
Looks like one of the cases picked up a nice bug, another one of the cases I'm not really sure that the ineffassign tool got it right. I didn't look closely at the CRI changes.

aanm reviewed (changes requested):
Please split the CRI related changes to be in a new PR so we can push it to 1.1

vadorovsky reviewed (commented):
@aanm @joestringer Separate pull requests created:
https://github.com/cilium/cilium/pull/4385
https://github.com/cilium/cilium/pull/4386

joestringer commented on pkg/bpf/bpffs.go:
```diff
@@ -258,7 +258,7 @@ func checkOrMountDefaultLocations() error {
 		}
 		if !cMounted {
-			err = mountFS()
+			err = mountFS() // ineffassign: ignore
```
Do you think that this ineffassign is because of the `... err := isBPFFS()` on line 256?

Maybe that line should be updated to not declare `err`. In that case, this may also be a bug and so the fix should be split into a separate PR and we can make sure it goes into 1.1.

raybejjani replied:
That might be it. Trying this out https://play.golang.org/p/S8liQ6Y4UsM without the if produced output as expected. I had thought assignments like this didn't shadow already existing variables unless it was in the `if` setup section (`if z=1; z < 9`).

I think I would prefer to refactor this function a little. Either use a big switch where each case corresponds to the various combinations of `mounted` and `bpffsInstance` or return early when possible. So, an early return on line 231, another early return when `mounted && bpffsInstance`, the implied default here. I'm not sure what the precedence is between `!mounted` and `!bpffsInstance` is  but both of those could return internally directly, since they are simple enough.

vadorovsky replied:
Yes, you are right. I also didn't expect that `v, err := something()` shadows the previously declared `err` variable, but it seems tat it does. I will early return in the separate PR.

joestringer reviewed (commented):
Looks good, one comment to follow up below. Also, could you update the developer documentation to describe how to install ineffassign since it'll be a build-time dependency?

raybejjani commented on pkg/kvstore/etcd.go:
```diff
@@ -580,7 +580,7 @@ reList:
 
 					// Retrieve latest revision
-					lastRev = 0
+					lastRev = 0 // ineffassign: ignore
```
Could we just move this line and/or the declaration of lastRev to the beginning of the `reList` loop? That's the only way we would end up with lastRev=0 anyway... I might have missed something, this code is incredibly difficult to follow.
While supporting ignore is handy, it defeats the point of using ineffassign. The tool is telling us that we've done something unintended in the language so our code is very likely incorrect.

vadorovsky replied:
You're right, I will try that and submit a separate PR if moving that assignment to the beginning of `reList` will work

raybejjani replied:
Thanks. To be clear, I'm not opposed to this PR being merged, but in the long run I'm worried the ignore will allow some bad behaviour on our part ("Oh, I just need to merge this to make this deadline" and then forgetting about it).

vadorovsky committed:
endpoint: Fix ineffassign error

`buildSuccess` is always going to receive its value later, so it
just needs to be declared.

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
kvstore: Fix inefassign error

`err` modification wasn't available outside of if statement scope.

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
option: Fix ineffassign error

´defaultValue` parameter was ineffective and didn't have any
impact on returned value.

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
test: Fix ineffassign errors

Always check `err` variables, ignore empty return values.

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
make: Add ineffassign as a part of precheck

Fixes #3609

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
vagrant: Add ineffassign tool

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

joestringer commented on pkg/kvstore/etcd.go:
```diff
@@ -298,7 +298,10 @@ func newEtcdClient(config *client.Config, cfgPath string) (BackendOperations, er
 			c.Close()
-			err = fmt.Errorf("unable to create default lease: %s", err)
+			if err = fmt.Errorf("unable to create default lease: %s", err); err != nil {
+				log.Fatal(err)
```
These lines could be de-duplicated by changing the if statement instead to:

```
err = ec.renewLease()
if err != nil {
...
```

This way the existing `log.Fatal()` would be hit, like the previous code intended.

joestringer commented on test/k8sT/Policies.go:
```diff
@@ -661,6 +661,7 @@ var _ = Describe("K8sPolicyTest", func() {
 			Expect(err).To(BeNil(), "Cannot get web pods")
 
 			serviceIP, port, err := kubectl.GetServiceHostPort(helpers.DefaultNamespace, "redis-master")
+			Expect(err).To(BeNil())
```
Please add a string comment to describe the problem here, similar to how line 661 has.

joestringer commented on test/runtime/chaos.go:
```diff
@@ -114,6 +114,7 @@ var _ = Describe("RuntimeChaos", func() {
 		status.ExpectFail("leftover interface were not properly cleaned up")
 
 		links, err := vm.Exec("sudo ip link show | wc -l").IntOutput()
+		Expect(err).Should(BeNil())
```
Again, please add an error string:

```
Expect(err).To(BeNil(), "Unable to list IPs")
```

joestringer commented on test/runtime/connectivity.go:
```diff
@@ -263,6 +263,7 @@ var _ = Describe("RuntimeConnectivityTest", func() {
 			err = helpers.RenderTemplateToFile(policyFileName, policy, os.ModePerm)
 			Expect(err).Should(BeNil())
 			_, err = vm.PolicyImportAndWait(helpers.GetFilePath(policyFileName), helpers.HelperTimeout)
+			Expect(err).Should(BeNil())
```
Please add error string.

joestringer reviewed (commented):
Some of the ineffassigns that are showing up in the testsuite show some really useful cases where we're unintentionally neglecting the error conditions. Looking forward to getting this in!

Please just add some error strings to the new expectations, it's difficult to debug what goes wrong when the test fails and doesn't have any string associated with the error.

joestringer reviewed (commented):
Just need to figure out why the CI "precheck" stage doesn't have ineffassign available and make it available.

joestringer reviewed (approved):
Just need to rebase against master when the builder-image has been updated to include `ineffassign`.

vadorovsky committed:
Dockerfile: Update cilium-builder to 2018-08-17

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
test: Update cilium-builder to 2018-08-17

Signed-off-by: Michal Rostecki <mrostecki@suse.de>
